noinline.go

```diff
@@ -0,0 +1,50 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/google/pprof/profile"
+)
+
+const grpcProcessDataFunc = "google.golang.org/grpc/internal/transport.(*loopyWriter).processData"
+
+// ApplyNoInlineHack removes problematic samples from the profile to avoid bad
+// inlining decisions that can have a large memory impact.
+// See https://github.com/golang/go/issues/65532 for details.
+//
+// TODO: Delete this once it's fixed upstream.
+func ApplyNoInlineHack(prof *profile.Profile) error {
+	if err := removeLeafSamples(prof, []string{grpcProcessDataFunc}); err != nil {
+		return fmt.Errorf("noinline hack: %w", err)
+	}
+	return nil
+}
+
+func removeLeafSamples(prof *profile.Profile, funcs []string) error {
+	newSamples := make([]*profile.Sample, 0, len(prof.Sample))
+	for _, s := range prof.Sample {
+		leaf, ok := leafLine(s)
+		if ok && lineContainsAny(leaf, funcs) {
+			continue
```

Comment:
The compiler looks at a CDF of sample weights to determine the cut-off of what is hot. By removing samples entirely, you are (ever so slightly) adjusting the overall weights, and thus might slightly impact other inlining decisions (in this case, I think you could get extra functions right on the edge of being considered "hot" move from "cold" to "hot").

Rather than removing samples entirely, I'd recommend mangling the symbol name so they don't match. Make them "UNKNOWN", or perhaps better yet "DO NOT INLINE: google.golang.org/grpc/internal/transport.(*loopyWriter).processData"

Reply:
Thank you Michael! I'll make that change <3.

Reply:
Fixed in https://github.com/DataDog/datadog-pgo/pull/4.